### PR TITLE
Policy gradient v2

### PR DESCRIPTION
**Pull request by gallupBenRyan**
Changes:
`config.cfg`: 
- added parameter to switch the Generator encoder from "LSTM" to "CNN" or back

`parser.py`: 
- added conditional code to keep original encoder as LSTM or as CNN, depending on config. 
- also added BatchNorm1d() to CNN encoder implementation
- changed `parse()` function to optionally take as input an incomplete Hypothesis and the generator states history and complete the parse (for use in the policy gradient rollout algorithm). `parse()` also optionally returns the Generator state history from generating a parse tree sequence, for when these need to be fed to the rollout procedure.
- added a `sample()` function to wrap around the rollout calls to `parse()`.
- Fixed `pretrain()` and `pgtrain()` functions so that they properly set the model to `train` state, and `pgtrain` now takes as an argument the states history for calculating rollout rewards.

`querygan_pyt.py`:
- `generate_samples()` function updated to return generated states for policy gradient rollout procedure when `oracle==True`
- 'generate_samples()` also updated with better progress bar reporting.
- Call to `rollout()` updated in line with changes to `rollout.py`
- Also fixed writing out of generator and discriminator loss histories.

`rollout.py`:
- Consolidated required `tree_rollout.py` functions into this file
- Added function `ffwd_hyp()` to create incomplete versions of generate hypotheses for rollout procedure.
- Fixed and cleaned up the main rollout iteration.

`vocab.py`:
- Added functionality to allow more versatile usage across projects.

### CODE REVIEW

File: src/gan/gen_pyt/model/parser.py [hunk 1]
```diff
@@ -889,10 +893,51 @@ def parse(self, src_sent, context=None, beam_size=5, debug=False):
...
+            src_sent: list of source utterance tokens
+            hyp: incomplete hypothesis
+            states: history of decoder states for generating Hypothesis
+            beam_size: beam size
+
+        Returns:
+            A completed Hypothesis
+        """
+
+        if hyp is not None and hyp.completed:
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/model/parser.py]: I'm not following what's going on here. Why is the double statement necessary? And, below, what's the purpose of the `while True`... what happens that's different the second time the `while` loop is executed that didn't happen the first time (or third, or fourth, etc.)?

File: src/gan/querygan_pyt.py [hunk 2]
```diff
@@ -142,8 +161,10 @@ def generate_samples(net, seq_len, generated_num, parser, oracle=False,
                 parfile.write(' '.join([str(x) for x in parents]) + '\n')
                 tokfile.write(' '.join(tokens) + '\n')
                 catfile.write('0' + '\n')
-    
-    return zip(samples, examples)
+
+        return None
```
Comment (mhoover-gallup) [inline src/gan/querygan_pyt.py]: Again, this concept of returning a different number of arguments is not best practice.

File: src/gan/utilities/rollout.py [hunk 3]
```diff
@@ -6,19 +6,18 @@
 import torch
 import torch.nn as nn
 
-from discriminator import MULTIVACDataset
+from discriminator import MULTIVACDataset, Tree
 from gen_pyt.asdl.lang.eng.eng_asdl_helper import asdl_ast_to_english
 from gen_pyt.model.parser import Parser
-from .tree_rollout import rollout_samples
+# from .tree_rollout import rollout_samples
```
Comment (mhoover-gallup) [inline src/gan/utilities/rollout.py]: Delete or uncomment

File: src/gan/utilities/rollout.py [hunk 4]
```diff
@@ -27,42 +26,97 @@ def hyp_to_parse(self, hyp, vocab):
...
-        tree = MULTIVACDataset.read_tree(parents)
+        parse = self.parser.get_parse(text)['sentences']
 
-        return tree, torch.tensor(vocab.convertToIdx(tokens, '<unk>'), 
-                                  dtype=torch.long, device='cpu')
-
-    # Need to figure out if the structure needs to change with the new
-    # models, or if this just works as is because they're sequences?
-    def get_tree_reward(self, examples, hyps, netD, vocab, verbose=False):
+        if len(parse) > 0:
```
Comment (mhoover-gallup) [inline src/gan/utilities/rollout.py]: Why are you only taking and parsing the first element of `parse`?

File: src/gan/utilities/rollout.py [hunk 5]
```diff
@@ -27,42 +26,97 @@ def hyp_to_parse(self, hyp, vocab):
...
-            # results is list of lists, (max_action_len-1, len(hyps))
-            samples = rollout_samples(self.ori_net, src_sents, hyps)
-            if verbose: print("Samples generated of shape ({},{})".format(len(samples), len(samples[0])))
 
-            inputs = [[]] * len(samples)
-
-            for x in tqdm(range(len(samples)), "Translating trees to discriminator..."):
-                for hyp in samples[x]:
-                    inputs[x].append(self.hyp_to_parse(hyp, vocab))
+            samples = [[0] * batch_size] * max_action_len
```
Comment (mhoover-gallup) [inline src/gan/utilities/rollout.py]: `[list(np.zeros(batch_size))] * max_action_len` might be more Pythonic.

File: src/gan/utilities/rollout.py [hunk 6]
```diff
@@ -27,42 +26,97 @@ def hyp_to_parse(self, hyp, vocab):
...
+            inputs  = [[0] * batch_size] * max_action_len
+            texts   = [[0] * batch_size] * max_action_len
+
+            for j in tqdm(range(1, max_action_len)):
```
Comment (mhoover-gallup) [inline src/gan/utilities/rollout.py]: What is `tqdm`?

**Comment by gallupBenRyan** [comment]
Superseded by #52, closing.
